docs/docs/admin-settings.md
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Platform Preferences

Platform wide settings can be configured using `Preferences` in Frontier. These preferences are stored in the database and are applied to all the users of the platform. Since these settings are crucial to the platform, only the platform admin can set these settings. In case the preferences are not set, the default values are used.

The settings can be set using the preferences CLI or the API as described below.

Note: The Frontier backend predefines some settings which are used by the platform. These settings are defined in the default traits of Frontier. The platform admin can override these settings by setting the same setting name using the preferences CLI or the API.

For example, the **disable_orgs_on_create** setting is defined in the default traits of Frontier. This setting is used by the platform to make the default state of new organizations in Frontier to be disabled. The platform admin can override this setting by setting the same setting name using the preferences CLI or the API.

## Usage

### Create and Update Preferences

<Tabs groupId="cli">
<TabItem value="API" label="API">

Use this [Create Platform Preference API](https://raystack-frontier.vercel.app/apis/admin-service-create-preferences) to create platform wide preferences.

```bash
$ curl -L -X POST 'http://127.0.0.1:7400/v1beta1/preferences' \
-H 'Content-Type: application/json' \
-H 'Accept: application/json' \
-H 'Authorization: Basic dGVzdFVzZXJuYW1lOnRlc3RQYXNzd29yZA==' \
--data-raw '{
  "preferences": [
    {
      "name": "disable_orgs_on_create",
      "value": "true"
    }
  ]
}'
```

</TabItem>
  <TabItem value="CLI" label="CLI" default>

```bash
## Syntax
$ frontier preferences set -n <name> -v <value>
$ frontier preferences set --name <name> --value <value>

## Example
$ frontier preferences set -n disable_orgs_on_create -v true -H X-Frontier-Email:user@raystack.org
```

</TabItem>
</Tabs>

Note: The setting name is case sensitive.

### List Preferences

Once the preferences are set, they can be listed using the following command:

<Tabs groupId="cli">
<TabItem value="API" label="API">

```bash
$ curl -L -X GET 'http://127.0.0.1:7400/v1beta1/preferences' \
-H 'Accept: application/json' \
-H 'Authorization: Basic dGVzdFVzZXJuYW1lOnRlc3RQYXNzd29yZA=='
```

</TabItem>

<TabItem value="CLI" label="CLI" default>

```bash
$ frontier preferences list
```

</TabItem>
</Tabs>

### Describe Preferences

The Preference traits are defined in the Frontier backend to be used by the platform. Only these traits can be used to set the platform wide preferences. The platform admin can view the default traits defined in Frontier using the following command:

<Tabs groupId="cli">
<TabItem value="API" label="API">

Use this [API](https://raystack-frontier.vercel.app/apis/frontier-service-describe-preferences) to list the default traits defined in Frontier, the traits with `app/platform` namespace denotes the platform wide traits.

```bash
$ curl -L -X GET 'http://127.0.0.1:7400/v1beta1/preferences/traits' \
-H 'Accept: application/json' \
-H 'Authorization: Basic dGVzdFVzZXI6dGVzdFBhc3M='
```

</TabItem>
  <TabItem value="CLI" label="CLI" default>

```bash
 $  frontier preferences get
```

</TabItem>
</Tabs>

This command will list all the default traits defined in Frontier. The traits with `app/platform` namespace denotes the platform wide traits. Similarly, the traits with `app/user`, `app/organization` and `app/group` namespace denotes the user, organization and group wide traits respectively.

## Platform Traits

All the trait description, default value and usage can be found with the above command. Just a brief description of the platform wide traits is given below. The input hints show the kind of values that can be provided for the trait separated by a comma. The default value is shown in bold.

### Usage and Description of Frontier Traits

| Name                         | Namespace    | Description                                                                                                                                                                       | Input Hints      |
| ---------------------------- | ------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------- |
| disable_orgs_on_create       | app/platform | If selected the new orgs created by members will be disabled by default.This can be used to prevent users from accessing the org until they contact the admin and get it enabled. | "**true**,false" |
| disable_orgs_listing         | app/platform | If selected will disallow non-admin APIs to list all organizations on the platform.                                                                                               | "true,**false**" |
| disable_users_listing        | app/platform | If selected will will disallow non-admin APIs to list all users on the platform                                                                                                   | "true,**false**" |
| invite_with_roles            | app/platform | Allow inviting new members with set of role ids. When the invitation is accepted, the user will be added to the org with the roles specified.                                     | "**true**,false" |
| invite_mail_template_subject | app/platform | The subject of the invite mail sent to new members                                                                                                                                |                  |
| invite_mail_template_body    | app/platform | The body of the invite mail sent to new members                                                                                                                                   |                  |
---

docs/docs/configurations.md
# Configuration

Frontier binary contains both the CLI client and the server. Each has it's own configuration in order to run. Server configuration contains information such as database credentials, spicedb connection, log severity, etc. while CLI client configuration only has configuration about which server to connect.

## Server Setup

There are several approaches to setup Frontier Server

1. [Using the CLI](#using-the-cli)
2. [Using the Docker](#using-the-docker)
3. [Using the Helm Chart](#using-the-helm-chart)

#### General pre-requisites

- PostgreSQL (version 13 or above)
- [SpiceDB](https://authzed.com/docs/spicedb/installing)

## Using the CLI

### Using config file

Create a config file with the following command

```bash
$ frontier server init
```

Alternatively you can [use `--config` flag](#using---config-flag) to customize to config file location.You can also [use environment variables](#using-environment-variable) to provide the server configuration.

Setup up the Postgres database, and SpiceDB instance and provide the details as shown in the example below.

> If you're new to YAML and want to learn more, see [Learn YAML in Y minutes.](https://learnxinyminutes.com/docs/yaml/)

Following is a sample server configuration yaml:

<details>
<summary> config.yaml </summary>

```yaml title=config.yaml
version: 1

# logging configuration
log:
  # debug, info, warning, error, fatal - default 'info'
  level: debug
  #  none(default), stdout, db
  audit_events: none

app:
  port: 8000
  grpc:
    port: 8001
    # optional tls configuration for grpc server
    tls_cert_file: "temp/server-cert.pem"
    tls_key_file: "temp/server-key.pem"
    tls_client_ca_file: "temp/ca-cert.pem"
  metrics_port: 9000
  # WARNING: identity_proxy_header bypass all authorization checks and shouldn't be used in production
  identity_proxy_header: X-Frontier-Email
  # full path prefixed with scheme where resources config yaml files are kept
  # e.g.:
  # local storage file "file:///tmp/resources_config"
  # GCS Bucket "gs://frontier/resources_config"
  resources_config_path: file:///tmp/resources_config\
  # secret required to access resources config
  # e.g.:
  # system environment variable "env://TEST_RULESET_SECRET"
  # local file "file:///opt/auth.json"
  # secret string "val://user:password"
  # optional
  resources_config_path_secret: env://TEST_RESOURCE_CONFIG_SECRET
  # disable_orgs_on_create if set to true will set the org status to disabled on creation. This can be used to
  # prevent users from accessing the org until they contact the admin and get it enabled. Default is false
  disable_orgs_on_create: false
  # disable_orgs_listing if set to true will disallow non-admin APIs to list all organizations
  disable_orgs_listing: false
  # disable_orgs_listing if set to true will disallow non-admin APIs to list all users
  disable_users_listing: false
  
  # configs for user invitation to join an organization
  invite:
    # with_roles if set to true will allow people in org with the permission to send invitation to users
    # with set of role ids. When the invitation is accepted, the user will be added to the org with the roles specified
    # This can be a security risk if the user who is inviting is not careful about the roles he is adding
    # and cause permission escalation
    # Note: this is dangerous and should be used with caution
    with_roles: false
    # invite email template (if not specified, default template will be used)
    mail_template:
      subject: "You have been invited to join an organization"
      body: "<div>Hi {{.UserID}},</div><br><p>You have been invited to join an organization: {{.Organization}}. Login to your account to accept the invitation.</p><br><div>Thanks,<br>Team Frontier</div>"
  # cors_origin is origin value from where we want to allow cors
  cors_origin: ["http://localhost:3000"]
  # configuration to allow authentication in frontier
  authentication:
    # to use frontier as session store
    session:
      # both of them should be 32 chars long
      # hash helps identify if the value is tempered with
      hash_secret_key: "hash-secret-should-be-32-chars--"
      # block helps in encryption
      block_secret_key: "block-secret-should-be-32-chars-"
      # domain used for setting cookies, if not set defaults to request origin host
      domain: ""
    # once authenticated, server responds with a jwt with user context
    # this jwt works as a bearer access token for all APIs
    token:
      # generate key file via "./frontier server keygen"
      # if not specified, access tokens will be disabled
      # example: /opt/rsa
      rsa_path: ""
      # issuer claim to be added to the jwt
      iss: "http://localhost.frontier"
      # validity of the token
      validity: "1h"
    # Public facing host used for oidc redirect uri and mail link redirection
    # after user credentials are verified.
    # If frontier is exposed behind a proxy, this should set as proxy endpoint
    # e.g. http://localhost:7400/v1beta1/auth/callback
    # Only the first host is used for callback by default, if multiple hosts are provided
    # they can be used to override the callback host for specific strategies using query param
    callback_urls: ["http://localhost:8000/v1beta1/auth/callback"]
    # by default, after successful authentication(flow completes) no operation will be performed,
    # to apply redirection in case of browsers, provide a list of urls one of which will be used
    # after authentication where users will be redirected to.
    # this is optional
    authorized_redirect_urls: []
    # oidc auth server configs
    oidc_config:
      google:
        client_id: "xxxxx.apps.googleusercontent.com"
        client_secret: "xxxxx"
        issuer_url: "https://accounts.google.com"
        # validity of the verification duration
        validity: "10m"
    mail_otp:
      subject: "Frontier - Login Link"
      # body is a go template with `Otp` as a variable
      body: "Please copy/paste the OneTimePassword in login form.<h2>{{.Otp}}</h2>This code will expire in 10 minutes."
      validity: "1h"
    mail_link:
      subject: "Frontier Login - One time link"
      # body is a go template with `Otp` as a variable
      body: "Click on the following link or copy/paste the url in browser to login.<br><h2><a href='{{.Link}}' target='_blank'>Login</a></h2><br>Address: {{.Link}} <br>This link will expire in 15 minutes."
      validity: 15m
  # platform level administration
  admin:
    # Email list of users which needs to be converted as superusers
    # if the user is already present in the system, it is promoted to su
    # if not, a new account is created with provided email id and promoted to su.
    # UUIDs/slugs of existing users can also be provided instead of email ids
    # but in that case a new user will not be created.
    users: []
  # smtp configuration for sending emails
  mailer:
    smtp_host: smtp.example.com
    smtp_port: 587
    smtp_username: "username"
    smtp_password: "password"
    smtp_insecure: true
    headers:
      from: "username@acme.org"
db:
  driver: postgres
  url: postgres://frontier:@localhost:5432/frontier?sslmode=disable
  max_query_timeout: 500ms

spicedb:
  host: spicedb.localhost
  pre_shared_key: randomkey
  port: 50051
  # fully_consistent ensures APIs although slower than usual will result in responses always most consistent
  # suggested to keep it false for performance
  fully_consistent: false
```

</details>

See [configuration reference](./reference/configurations.md) for more details.

### Using environment variables

All the server configurations can be passed as environment variables using underscore \_ as the delimiter between nested keys.

<details>
<summary> .env </summary>

```bash
LOG_LEVEL=debug
APP_PORT=8000
APP_GRPC_PORT=8001
DB_DRIVER=postgres
DB_URL=postgres://frontier:@localhost:5432/frontier?sslmode=disable
DB_MAX_QUERY_TIMEOUT=500ms
SPICEDB_HOST=spicedb.localhost
SPICEDB_PRE_SHARED_KEY=randomkey
SPICEDB_PORT=50051
SPICEDB_FULLY_CONSISTENT=false
```

</details>

Set the env variable using export

```bash
$ export DB_PORT = 5432
```

### Starting the server

Database migration is required during the first server initialization. In addition, re-running the migration command might be needed in a new release to apply the new schema changes (if any). It's safer to always re-run the migration script before deploying/starting a new release.

To initialize the database schema, Run Migrations with the following command:

```bash
$ frontier server migrate
```

To run the Frontier server use command:

```bash
$ frontier server start
```

#### Using `--config` flag

```bash
$ frontier server migrate --config=<path-to-file>
```

```bash
$ frontier server start --config=<path-to-file>
```

## Using the Docker

To run the Frontier server using Docker, you need to have Docker installed on your system. You can find the installation instructions [here](https://docs.docker.com/get-docker/).

You can choose to set the configuration using environment variables or a config file. The environment variables will override the config file.

If you use Docker to build frontier, then configuring networking requires extra steps. Following is one of doing it by running postgres and spicedb inside with `docker-compose` first.

Go to the root of this project and run `docker-compose`.

```bash
$ docker-compose up
```

Once postgres and spicedb has been ready, we can run Frontier by passing in the config of postgres and elasticsearch defined in `docker-compose.yaml` file.

### Using config file

Alternatively you can use the `frontier.yaml` config file defined [above](#using-config-file) and run the following command.

```bash
$ docker run -d \
    --restart=always \
    -p 7400:7400 \
    -v $(pwd)/frontier.yaml:/frontier.yaml \
    --name frontier-server \
    raystack/frontier:<version> \
    server start -c /config.yaml
```

### Using environment variables

All the configs can be passed as environment variables using underscore `_` as the delimiter between nested keys. See the example as discussed [above](#using-environment-variable)

Run the following command to start the server

```bash
$ docker run -d \
    --restart=always \
    -p 7400:7400 \
    --env-file .env \
    --name frontier-server \
    raystack/frontier:<version> \
    server start
```

## Using the Helm chart

### Pre-requisites for Helm chart

Frontier can be installed in Kubernetes using the Helm chart from https://github.com/raystack/charts.

Ensure that the following requirements are met:

- Kubernetes 1.14+
- Helm version 3.x is [installed](https://helm.sh/docs/intro/install/)

### Add Raystack Helm repository

Add Raystack chart repository to Helm:

```bash
helm repo add raystack https://raystack.github.io/charts/
```

You can update the chart repository by running:

```bash
helm repo update
```

### Setup helm values

The following table lists the configurable parameters of the Frontier chart and their default values.

See full helm values guide [here](https://github.com/raystack/charts/tree/main/stable/frontier#values) and [values.yaml](https://github.com/raystack/charts/blob/main/stable/frontier/values.yaml) file

Install it with the helm command line:

```bash
helm install my-release -f values.yaml raystack/frontier
```

## Client Initialisation

Add a client configurations file with the following command:

```bash
frontier config init
```

Open the config file and edit the gRPC host for Frontier CLI

```yml title="frontier.yaml"
client:
  host: localhost:8081
```

List the client configurations with the following command:

```bash
frontier config list
```

#### Required Header/Metadata in API

In the current version, all HTTP & gRPC APIs in Frontier requires an identity header/metadata in the request. The header key is configurable but the default name is `X-Frontier-Email`.

If everything goes well, you should see something like this:

```bash
2023-05-17T00:02:54.324+0530    info    frontier starting {"version": "v0.5.1"}
2023-05-17T00:02:54.331+0530    debug   resource config cache refreshed {"resource_config_count": 0}
2023-05-17T00:02:54.333+0530    info    Connected to spiceDB: localhost:50051
2023-05-17T00:02:54.339+0530    info    metaschemas loaded      {"count": 4}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.5-rc.5](https://github.com/odpf/shield/compare/v0.1.5-rc.4...v0.1.5-rc.5) (2021-11-09)


### Features

* **proxy:** add encodeParams to encode parameter variables ([98057d2](https://github.com/odpf/shield/commit/98057d2ff04f4595bbbb78fc44cce436d1a48834))
* **proxy:** put encodeParams inside extraOptions ([baa256d](https://github.com/odpf/shield/commit/baa256d66d61eb4a35f91d6c10e8a9c839d5e1f0))

### [0.1.5-rc.4](https://github.com/odpf/shield/compare/v0.1.5-rc.3...v0.1.5-rc.4) (2021-08-19)

### [0.1.5-rc.3](https://github.com/odpf/shield/compare/v0.1.5-rc.1...v0.1.5-rc.3) (2021-08-17)

### [0.1.5-rc.2](https://github.com/odpf/shield/compare/v0.1.5-rc.1...v0.1.5-rc.2) (2021-08-17)

### [0.1.5-rc.1](https://github.com/odpf/shield/compare/v0.1.5-rc.0...v0.1.5-rc.1) (2021-08-13)


### Features

* add redis caching for user groups api ([70cefd2](https://github.com/odpf/shield/commit/70cefd297fbc68e17390fc3bba37dde74c75b96e))

### [0.1.5-rc.0](https://github.com/odpf/shield/compare/v0.1.4...v0.1.5-rc.0) (2021-08-13)


### Features

* add redis caching for group list api ([d53b7d6](https://github.com/odpf/shield/commit/d53b7d657cf432831457e6c37b31c037b9e23818))


### Bug Fixes

* logs for http request ([b2a9fc6](https://github.com/odpf/shield/commit/b2a9fc6bf94ec30cc865c24be3e1ba86f6f338c0))

### [0.1.4](https://github.com/odpf/shield/compare/v0.1.3...v0.1.4) (2021-08-09)


### Bug Fixes

* check for more than 0 role tags for filter ([4ae69b1](https://github.com/odpf/shield/commit/4ae69b15eecfedf86cdd64f03f8d2de0b1c08b43))

### [0.1.3](https://github.com/odpf/shield/compare/v0.1.2...v0.1.3) (2021-08-06)


### Bug Fixes

* typo in conf file destination postbuild ([89c455c](https://github.com/odpf/shield/commit/89c455c0618f844c2431b2a9fafbc0f5a85b0d24))

### [0.1.2](https://github.com/odpf/shield/compare/v0.1.1...v0.1.2) (2021-08-06)


### Bug Fixes

* build folder structure ([ea3f9cc](https://github.com/odpf/shield/commit/ea3f9cce164a71a0b4866b6de818b89cc9af9be0))
* policy filter in get all user api ([#16](https://github.com/odpf/shield/issues/16)) ([37e6911](https://github.com/odpf/shield/commit/37e69115942c2c0284aa5fec5559a3edafc9c6f2))

### [0.1.1](https://github.com/odpf/shield/compare/v0.1.0...v0.1.1) (2021-07-12)


### Features

* add tags filtering for roles ([#10](https://github.com/odpf/shield/issues/10)) ([85ca9fa](https://github.com/odpf/shield/commit/85ca9fa43c7b1fa9ff2737dcd575daad2c02b837))

## 0.1.0 (2021-05-24)


### Features

* add default role for when users are added to group ([8a63795](https://github.com/odpf/shield/commit/8a63795d0f5ae27ea6b89133fe3ac30389a9d910))
* add memberCount to get group by id API ([88a859d](https://github.com/odpf/shield/commit/88a859d08fae4ec443876442bd0c34090d7d6e7d))
* add newrelic for monitoring ([0a8e6b7](https://github.com/odpf/shield/commit/0a8e6b7d371e583394936da432e59394db6bf9c2))
* add resource attributes mapping create/delete API ([f6569a5](https://github.com/odpf/shield/commit/f6569a578d393473a98c0bc0c97aabef3156c1b0))
* add role response to update role API ([2832e26](https://github.com/odpf/shield/commit/2832e264814b2bb63810239ca275c313e491d3ac))
* add test github action ([b100853](https://github.com/odpf/shield/commit/b100853e8f4d54a234be6b3e0fdc953ac4209752))
* add tests for both loadFiltered and loadAll policies ([7642680](https://github.com/odpf/shield/commit/76426809e74a5b17c29cb48dc3cdf7291436d57b))
* add unknown for response validation ([a1f1390](https://github.com/odpf/shield/commit/a1f13904108602fecfeadc55f8424eeedef4f8a2))
* add user-id in reverse proxy headers ([10839ca](https://github.com/odpf/shield/commit/10839cafb5dcdba785bf3328acb250003c1ff0bb))
* add username and groupname columns ([5f7ffb0](https://github.com/odpf/shield/commit/5f7ffb0d0cfccf1ea13ea70b0cb5ee91eab7846a))
* add validation for username and groupname ([16e092b](https://github.com/odpf/shield/commit/16e092bd17a0e159db28273e62849d96672eef6c))
* allow unknown query parameters ([2c7e357](https://github.com/odpf/shield/commit/2c7e35772cb4972fdee70a39a650a716da807fb0))
* delete generic model subscriber, add group subscribe for group related event ([0278661](https://github.com/odpf/shield/commit/0278661476c82eaa0cb69afa13174fca41fe0372))
* fix broken test cases ([d89b571](https://github.com/odpf/shield/commit/d89b571764981a436f06ec8296f13d3e9672d3d2))
* further optimize policy subset loading ([d4087b6](https://github.com/odpf/shield/commit/d4087b656b9c15866a2c422d0b6b56da30b39050))
* get activities based on group , test cases for activity handler and resources ([e6d6984](https://github.com/odpf/shield/commit/e6d6984dce6854790e3cdb9ccb241de3016e5356))
* get all activity log or based on group name ([a0f4108](https://github.com/odpf/shield/commit/a0f41082b9952b85062501fdda2d263beeefbf25))
* get list of groups based on attribute filters ([15fa751](https://github.com/odpf/shield/commit/15fa7510ed2f3f6d49c26ff8450735c37b043460))
* get list of groups of a user ([9852001](https://github.com/odpf/shield/commit/9852001de1d4afc6e16d0d571980faed6d067382))
* get list of users ([2960125](https://github.com/odpf/shield/commit/29601255c94eab7a32c09467ca2c0dcd9d665f8f))
* get users of a group with policies ([0a318e4](https://github.com/odpf/shield/commit/0a318e4a6ca2908e4f1fd3c3e5c81a4982fd7bef))
* get users of a group with policies with policy filters ([b0568c3](https://github.com/odpf/shield/commit/b0568c3261d685d24939332b8aada1171f0b472d))
* group proxy related api in swagger ([c21bde4](https://github.com/odpf/shield/commit/c21bde4b6b4df3730886c667e684724ffd7c713a))
* implement ability to load subset of a policy ([7d0c456](https://github.com/odpf/shield/commit/7d0c45671a0e162e0e5e35f8af3e738a68c4f278))

pkg/server/config.go
package server

import (
	"fmt"

	"github.com/raystack/frontier/pkg/server/interceptors"

	"github.com/raystack/frontier/pkg/mailer"

	"github.com/raystack/frontier/internal/bootstrap"

	"github.com/raystack/frontier/core/authenticate"
	"github.com/raystack/frontier/pkg/telemetry"
)

type GRPCConfig struct {
	Port            int    `mapstructure:"port" default:"8081"`
	MaxRecvMsgSize  int    `mapstructure:"max_recv_msg_size" default:"33554432"`
	MaxSendMsgSize  int    `mapstructure:"max_send_msg_size" default:"33554432"`
	TLSCertFile     string `mapstructure:"tls_cert_file" default:""`
	TLSKeyFile      string `mapstructure:"tls_key_file" default:""`
	TLSClientCAFile string `mapstructure:"tls_client_ca_file" default:""`
}

func (cfg Config) grpcAddr() string { return fmt.Sprintf("%s:%d", cfg.Host, cfg.GRPC.Port) }

type Config struct {
	// port to listen HTTP requests on
	Port int `yaml:"port" mapstructure:"port" default:"8080"`

	// GRPC Config
	GRPC GRPCConfig `mapstructure:"grpc"`

	// metrics port
	MetricsPort int `yaml:"metrics_port" mapstructure:"metrics_port" default:"9000"`

	// the network interface to listen on
	Host string `yaml:"host" mapstructure:"host" default:"127.0.0.1"`

	// TODO might not suitable here because it is also being used by proxy
	// Headers which will have user's email id
	IdentityProxyHeader string `yaml:"identity_proxy_header" mapstructure:"identity_proxy_header" default:""`

	// ResourcesPath is a directory path where resources is defined
	// that this service should implement
	ResourcesConfigPath string `yaml:"resources_config_path" mapstructure:"resources_config_path"`

	// ResourcesPathSecretSecret could be an env name, file path or actual value required
	// to access ResourcesPathSecretPath files
	ResourcesConfigPathSecret string `yaml:"resources_config_path_secret" mapstructure:"resources_config_path_secret"`

	TelemetryConfig telemetry.Config `yaml:"telemetry_config" mapstructure:"telemetry_config"`

	Authentication authenticate.Config `yaml:"authentication" mapstructure:"authentication"`

	// Deprecated: use Cors instead
	CorsOrigin []string `yaml:"cors_origin" mapstructure:"cors_origin"`
	// Cors configuration setup origin value from where we want to allow cors
	// headers and methods are the list of headers and methods we want to allow
	Cors interceptors.CorsConfig `yaml:"cors" mapstructure:"cors"`

	Admin bootstrap.AdminConfig `yaml:"admin" mapstructure:"admin"`

	Mailer mailer.Config `yaml:"mailer" mapstructure:"mailer"`
}
